### ISSUE TITLE
rclone Google Photos cannot sync the data before 2000

### ISSUE THREAD

**Issue by YNZone** [rclone/rclone#3630]
#### Output of `rclone version`

rclone v1.49.5

#### Describe the issue

rclone Google Photos cannot sync the data before 2000，but gphotos-sync(https://pypi.org/project/gphotos-sync/) can do this.

**Comment by ncw** [comment]
You should be able to read the media in media/all and in album.

It is quite expensive finding the oldest file in the photos album so rclone hard codes the 2000 value.

This could be a parameter? What do you think?

And the reply on the forum was:

> I synced only media/all and album.
> I don't know where to look for photos data before 2000（year）.

**Comment by ncw** [comment]
`media/all` should get you all the photos regardless of age.

This could be made configurable

https://github.com/rclone/rclone/blob/38652d046de40ce169a83caddef2f96a9ebd6005/backend/googlephotos/pattern.go#L229

**Comment by YNZone** [comment]
i think made it configurable maybe better… thanks

**Comment by ncw** [comment]
Do you want to try doing this?  I can talk you through it.

**Comment by YNZone** [comment]
Yes！Please tell me the test method.

**Comment by sakshamkhanna** [comment]
It doesn't feel right, having this in  a config param. 
While setting up a connection, the user shouldn't be expected to know the timestamp of the oldest photo in the library.

One way to do it could be in the googlephotos.init().
Right after registering the regInfo, make a bunch of api calls to GET media with DateFilter.Range param used like a binary search between 1800(??) and time.Now() to narrow down to the year in which the first media item appears

This seems a bit hacky, which leads to an idea for a new feature. 
A callback function that gets triggered everytime a remote is added/authenticated/refreshed to enable unique workflows like this. 

Any thoughts?

**Comment by ncw** [comment]
> It doesn't feel right, having this in a config param.

It isn't ideal!

> While setting up a connection, the user shouldn't be expected to know the timestamp of the oldest photo in the library.
> 
> One way to do it could be in the googlephotos.init().
> Right after registering the regInfo, make a bunch of api calls to GET media with DateFilter.Range param used like a binary search between 1800(??) and time.Now() to narrow down to the year in which the first media item appears

Binary search is a nice idea :-)

It will take about 8 queries to do this at 1-2 second each so a noticeable amount of time.

> This seems a bit hacky, which leads to an idea for a new feature.
> A callback function that gets triggered everytime a remote is added/authenticated/refreshed to enable unique workflows like this.

There is one already.  So this could run to update the value stored in the config file.

**Comment by ncw** [comment]
@YNZone wrote:
> Yes！Please tell me the test method.

First step create a new option here, called `start_year` type `int`, default 2000, Advanced: true

https://github.com/rclone/rclone/blob/22bf8589cdaaea6a2fa15a607f4fc2f330e16157/backend/googlephotos/googlephotos.go#L137

Add it to the options struct

https://github.com/rclone/rclone/blob/22bf8589cdaaea6a2fa15a607f4fc2f330e16157/backend/googlephotos/googlephotos.go#L142

Next step use that option here

https://github.com/rclone/rclone/blob/22bf8589cdaaea6a2fa15a607f4fc2f330e16157/backend/googlephotos/pattern.go#L227

You'll need to make `years` a method on `Fs` so it can read `f.opt.StartYear`

**Comment by YNZone** [comment]
Sorry, this is a bit difficult. I think it would be better to wait for the 1.51 version.

**Comment by sakshamkhanna** [comment]
> There is one already. So this could run to update the value stored in the config file.

@ncw could point me to where the callback function is configured. I cant seem to find it

**Comment by ncw** [comment]
This is the config callback

https://github.com/rclone/rclone/blob/18c029e0f0c3a5dd0c925e285ce39071cb8728ff/backend/googlephotos/googlephotos.go#L82

**Comment by darkLord19** [comment]
@ncw 
Hi I just came across this project and I am interested in contributing to the project and I find this issue a good start to get in. 
> You'll need to make `years` a method on `Fs` so it can read `f.opt.StartYear`

When you say make `years` a method on `Fs`, do you mean to make years a method of interface Fs defined here?
https://github.com/rclone/rclone/blob/ab895390f4b46f5455a508b1305a497026351a56/fs/fs.go#L254

**Comment by ncw** [comment]
> When you say make `years` a method on `Fs`, do you mean to make years a method of interface Fs defined here?
> https://github.com/rclone/rclone/blob/ab895390f4b46f5455a508b1305a497026351a56/fs/fs.go#L254

The function `years` needs to be a method on the type Fs defined in the backend

https://github.com/rclone/rclone/blob/1868c77e16b233b87fbc02f3b6a42c6f1f094ba6/backend/googlephotos/googlephotos.go#L148

so become

```go
 func (f *Fs) years(ctx context.Context, f lister, prefix string, match []string) (entries fs.DirEntries, err error) { 
```

Then instead of using the constant `2000` you can use `f.opt.StartYear`

https://github.com/rclone/rclone/blob/1868c77e16b233b87fbc02f3b6a42c6f1f094ba6/backend/googlephotos/pattern.go#L229

**Comment by dooven** [comment]
I'd like to tackle this one if no one has yet.

**Comment by ncw** [comment]
@dooven wrote:
> I'd like to tackle this one if no one has yet.
That would be great - thank you :-)

**Comment by ncw** [comment]
Thanks to @dooven this is fixed!

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52
